### ISSUE TITLE
Tools/scripts/patchcheck.py must ignore changes in Modules/expat/ and Modules/zlib/ subdirectories

### ISSUE THREAD

**Issue by vstinner** [python/cpython#75404]
BPO | [31221](https://bugs.python.org/issue31221)
--- | :---
Nosy | @vstinner, @skrah, @serhiy-storchaka
PRs | <li>python/cpython#3109</li><li>python/cpython#3116</li><li>python/cpython#3118</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = <Date 2017-08-17.15:13:35.047>
created_at = <Date 2017-08-16.15:59:09.985>
labels = ['3.7']
title = 'Tools/scripts/patchcheck.py must ignore changes in Modules/expat/ and Modules/zlib/ subdirectories'
updated_at = <Date 2017-08-17.15:13:35.045>
user = 'https://github.com/vstinner'
```

bugs.python.org fields:
```python
activity = <Date 2017-08-17.15:13:35.045>
actor = 'vstinner'
assignee = 'none'
closed = True
closed_date = <Date 2017-08-17.15:13:35.047>
closer = 'vstinner'
components = []
creation = <Date 2017-08-16.15:59:09.985>
creator = 'vstinner'
dependencies = []
files = []
hgrepos = []
issue_num = 31221
keywords = []
message_count = 11.0
messages = ['300372', '300380', '300410', '300413', '300419', '300426', '300427', '300428', '300430', '300435', '300436']
nosy_count = 3.0
nosy_names = ['vstinner', 'skrah', 'serhiy.storchaka']
pr_nums = ['3109', '3116', '3118']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue31221'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by vstinner** [comment]
My PR https://github.com/python/cpython/pull/3106 failed because patchcheck wants me to normalize spaces of [Modules/expat/siphash.h](https://github.com/python/cpython/blob/main/Modules/expat/siphash.h), whereas this file is copied from libexpat: see bpo-30947.

Attached PR changes [Tools/scripts/patchcheck.py](https://github.com/python/cpython/blob/main/Tools/scripts/patchcheck.py) to ignore changes in [Modules/expat/](https://github.com/python/cpython/blob/main/Modules/expat) and [Modules/zlib/](https://github.com/python/cpython/blob/main/Modules/zlib) subdirectories. The change also drops support for Mercurial, since CPython migrated to Git.

**Comment by serhiy-storchaka** [comment]
Is it worth to add other imported libs (libmpdec, libffi)? Or may be just remove trailing spaces when import sources from upstream?

**Comment by vstinner** [comment]
Serhiy: "Is it worth to add other imported libs (libmpdec, libffi)?"

Oh right, I forgot libmpdec: I updated my PR.

By the way, I started a list of embedded libraries:
http://haypo-notes.readthedocs.io/cpython.html#embedded-libraries

* zlib
* expat
* libmpdec
* libffi

libffi should be excluded in Python 2.7 and 3.6, but it was removed from Python 3.7.


Serhiy: "Or may be just remove trailing spaces when import sources from upstream?"

I prefer to leave the copied code unchanged, so it allows to easily cherry-pick from upstream. I did that once for libexpat (for fix a compilation issue on Visual Studio 2008).

**Comment by serhiy-storchaka** [comment]
> libffi should be excluded in Python 2.7 and 3.6, but it was removed from Python 3.7.

The parts for OSX and MSVC still are in the source tree.

As for dropping support for Mercurial, is patchcheck.py used in other Python implementations (PyPy, Jython) which still use Mercurial?

**Comment by 5531d0d8-2a9c-46ba-8b8b-ef76132a492c** [comment]
Yes, please add libmpdec to the ignored files.

**Comment by vstinner** [comment]
Serhiy: "The parts for OSX and MSVC still are in the source tree."

Oh, I didn't know that we still had two special libffi directories for macOS and Windows!

Why do we still have these directories? Both seems to be very outdated, no?

I completed my list:
http://haypo-notes.readthedocs.io/cpython.html#embedded-libraries


Serhiy: "As for dropping support for Mercurial, is patchcheck.py used in other Python implementations (PyPy, Jython) which still use Mercurial?"

I reverted my change to keep Mercurial support.

**Comment by vstinner** [comment]
Stefan Krah: "Yes, please add libmpdec to the ignored files."

Ok, (already) done.

**Comment by vstinner** [comment]
New changeset 4a347ce426fe7381885703d9074d7a6b3aeb2f2b by Victor Stinner in branch 'master':
bpo-31221: patchcheck ignores external libraries (bpo-3109)
https://github.com/python/cpython/commit/4a347ce426fe7381885703d9074d7a6b3aeb2f2b

**Comment by vstinner** [comment]
New changeset d23b1c4fd82a326e729027a791220d6011e097b4 by Victor Stinner in branch '3.6':
[3.6] bpo-31221: patchcheck ignores external libraries (bpo-3109) (bpo-3116)
https://github.com/python/cpython/commit/d23b1c4fd82a326e729027a791220d6011e097b4

**Comment by vstinner** [comment]
New changeset d45cb040b9fcd74e8824d417bf789fcef4c9592c by Victor Stinner in branch '2.7':
[2.7] bpo-31221: patchcheck ignores external libraries (bpo-3109) (bpo-3118)
https://github.com/python/cpython/commit/d45cb040b9fcd74e8824d417bf789fcef4c9592c

**Comment by vstinner** [comment]
patchcheck fixed in 2.7, 3.6 and 3.7 master.
